### PR TITLE
Convert _child_max to an iterative algorithm

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by cpcloud** [comment]
`ndependents` is also recursive and should be converted to an iterative version

File: dask/order.py [hunk 1]
```diff
@@ -149,18 +148,28 @@ def child_max(dependencies, dependents, scores):
...
-        result[key] = max([_child_max(k, scores, result, dependencies,
-                                      dependents) for k in deps]) - 0.5
-    return result[key]
+        for dep in deps:
+            stack.append((dep, result.setdefault(dep, value)))
+
+        # if key has deps, then its value is the max of all the values of its
+        # deps - 0.5
+        if deps:
+            result.setdefault(key, max(value for _, value in stack) - 0.5)
```
Comment (mrocklin) [inline dask/order.py]: This max over all values in the stack seems like it increases the algorithmic complexity of the algorithm beyond linear.  How does this fare on large graphs?

Comment (mrocklin) [inline dask/order.py]: The recursive version effectively memoized off of previously computed results

Comment (cpcloud) [inline dask/order.py]: I can simulate that part of the recursive version by checking if `key in result` rather than using `setdefault`

Comment (mrocklin) [inline dask/order.py]: Even in the case of a miss do you have to iterate over all elements in the stack?  If so how large is the stack?

Comment (cpcloud) [inline dask/order.py]: No, I can keep track of the max.

File: dask/order.py [hunk 2]
```diff
@@ -93,28 +94,28 @@ def ndependents(dependencies, dependents):
...
+        key, value = stack.pop()
         deps = dependents[key]
-        result[key] = sum(
-            [_ndependents(k, result, dependencies, dependents)
-             for k in deps]) + 1
-    return result[key]
+        ndeps = 0
+        for dep in deps:
+            val = result.setdefault(dep, value)
+            stack.append((dep, val))
```
Comment (mrocklin) [inline dask/order.py]: This still seems to add all dependents many times onto the stack.  I would expect there to be some check like 

```
if we haven't seen this dependency before:
    add it on to the stack
```

I may be wrong about this though.

File: dask/order.py [hunk 3]
```diff
@@ -93,28 +94,28 @@ def ndependents(dependencies, dependents):
...
-
-    result.update(dict((r, 1) for r in roots))
-
+    result = dict((k, 1) for k, v in dependents.items() if not v)
     leaves = [k for k, v in dependencies.items() if not v]
-
     for leaf in leaves:
-        _ndependents(leaf, result, dependencies, dependents)
-
+        _ndependents(leaf, result, dependents)
```
Comment (mrocklin) [inline dask/order.py]: I wonder if there is a way to do this once by dumping all of the leaves onto the stack and then iterating

**Comment by mrocklin** [comment]
Here is a fail case

```python
In [1]: from dask.order import *

In [2]: dsk = {'a': 1, 'b': (inc, 'a'), 'c': (inc, 'b'), 'd': (inc, 'c')}

In [3]: dependencies, dependents = get_deps(dsk)

In [4]: ndependents(dependencies, dependents)
Out[4]: {'a': 3, 'b': 3, 'c': 2, 'd': 1}
```

We can't accurately compute the number of dependents of `a` before we know the number of dependents of `b`.  I'm unsure of the `setdefault` trick.  

In general the traversal of stack-based "recursion" schemes should probably still reflect the original recursive flow.  In this case we add a bunch of stuff onto the (call) stack until we get to a base case, in this case one of the root nodes.  The first element that we add to results should be the root nodes followed by one of their children.

**Comment by mrocklin** [comment]
So we might start from the leaves and do one of the following

```python
while stack:
    key = stack.pop()
    deps = dependents[key]
    new = [d for d in deps if d not in result]  # TODO: check to ensure that this is `O(len(deps))`
    if not new:
        result[key] = sum(map(results.get, deps)) + 1
    else:
        stack.append(key)
        stack.extend(deps)
```

**Comment by mrocklin** [comment]
What is the status on this?  @cpcloud should I close this?  Is there something that we can salvage from here for the future (like a test that shows the issue?)

**Comment by cpcloud** [comment]
The test is passing with no other changes which suggests I either didn't reproduce it correctly or it's no longer an issue. I'm going to close this for now as my time is limited right now.

**Comment by mrocklin** [comment]
This just came up again by David Mertz (for whom I do not have a github handle).  Reopening.

**Comment by DavidMertz** [comment]
Thanks.  My specific code:

```python
df = dd.read_csv('/Users/dmertz/Data/NYC-Taxi/trip_data_*.csv',
                 skipinitialspace=True,
                 parse_dates=['pickup_datetime','dropoff_datetime'])
(df.set_index('pickup_datetime', compute=False)
   .to_castra('/Users/dmertz/Data/NYC-Taxi/trip_data.castra', categories=True))
```

Using `dask.diagnostics.ProgressBar`, I can see the `.set_index()` runs for about 12.5 min, then a second progress bar is created for the `.to_castra()`.  This latter fails almost immediately.

**Comment by DavidMertz** [comment]
Since the exception happens almost immediately on the `.to_castra()`, I wonder if a cheap fix might be to wrap to call with a `try/except` in a `while`, and just multiply the current recursion limit by 10x each time you see that particular problem.  You could reset it in the `finally` to avoid polluting the rest of the interpreter.

**Comment by eriknw** [comment]
We cannot guarantee that the Python interpreter will respect the recursion limit that we set.  In practice, the CPython interpreter (or at least the 2.x series of it) has an upper limit to what it will handle.  Converting recursive algorithms to iterative is probably the better option if there is a reasonable chance of reaching the default recursion limit.

**Comment by eriknw** [comment]
Is there anything I should use from this PR, or is it okay if I make a clean one?  I took a completely different approach than was done here at making `child_max` and `ndependents` iterative.

**Comment by mrocklin** [comment]
Clean seems fine with me

**Comment by DavidMertz** [comment]
@eriknw: I was thinking more of the quick-and-dirty temporary fix than the proper iterative re-factoring.  Anyway, for the NYC Taxi data, my specific increase of recursionlimit to 10000 indeed succeeded.

**Comment by mrocklin** [comment]
Closing this in favor of #744

**Pull request closed without merging**